File: src/cpp/sql_injection.cpp
// CWE-89: SQL injection — environment input concatenated into a query string.
#include <cstdlib>
#include <string>

extern int db_exec(const char* sql);   // pretend DB API

int main() {
    const char* userId = std::getenv("USER_ID");   // untrusted source
    std::string sql = "SELECT * FROM users WHERE id = '" +
                      std::string(userId ? userId : "") + "'";
    return db_exec(sql.c_str());                    // SQL sink
}

File: src/javascript/sql-injection.js
// CWE-89: SQL Injection
// User-controlled input is concatenated directly into a SQL statement.

const express = require('express');
const mysql = require('mysql');

const app = express();
const db = mysql.createConnection({
  host: 'localhost',
  user: 'root',
  password: 'hunter2',
  database: 'app',
});

app.get('/user', (req, res) => {
  const userId = req.query.id;
  const sql = "SELECT * FROM users WHERE id = '" + userId + "'";
  db.query(sql, (err, rows) => {
    if (err) return res.status(500).send(err.message);
    res.json(rows);
  });
});

app.get('/search', (req, res) => {
  const term = req.query.q;
  db.query(`SELECT * FROM products WHERE name LIKE '%${term}%'`, (err, rows) => {
    if (err) return res.status(500).send(err.message);
    res.json(rows);
  });
});

app.post('/login', express.urlencoded({ extended: true }), (req, res) => {
  const { username, password } = req.body;
  const sql = `SELECT * FROM users WHERE username = '${username}' AND password = '${password}'`;
  db.query(sql, (err, rows) => {
    if (err) return res.status(500).send(err.message);
    res.json({ ok: rows.length > 0 });
  });
});

app.listen(3000);
